tmail-backend/blob/blobid-list/src/main/java/com/linagora/tmail/blob/blobid/list/SingleSaveBlobStoreDAO.java

```diff
@@ -0,0 +1,43 @@
+package com.linagora.tmail.blob.blobid.list;
+
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.BlobStoreDAO;
+import org.apache.james.blob.api.BucketName;
+import org.apache.james.blob.api.ObjectStoreException;
+import reactor.core.publisher.Mono;
+
+import javax.inject.Inject;
+import java.util.Objects;
+
+public class SingleSaveBlobStoreDAO {
+    private final BlobStoreDAO blobStoreDAO;
+    private final BlobIdList blobIdList;
+    private final BucketName defaultBucketName;
+
+    @Inject
+    public SingleSaveBlobStoreDAO(BlobStoreDAO blobStoreDAO,
+                                  BlobIdList blobIdList,
+                                  BucketName defaultBucketName) {
+        this.blobStoreDAO = blobStoreDAO;
+        this.blobIdList = blobIdList;
+        this.defaultBucketName = defaultBucketName;
+    }
+
+    public Mono<Void> save(BucketName bucketName, BlobId blobId, byte[] data) {
+        return Mono.from(blobIdList.isStored(blobId))
+                .filter(isStored -> Objects.equals(isStored, false))
+                .switchIfEmpty(Mono.error(new ObjectStoreException("Can not save duplicate blob when single save is enabled")))
+                .then(Mono.from(blobStoreDAO.save(bucketName, blobId, data))
+                        .then(Mono.from(blobIdList.store(blobId))))
```

Comment:
Thought: Risk of concurrency issues here? If saving a big blob, the time it's being saved, we could still try to save an other one as the blobId has not been stored yet... 

I don't think we should save the blobId first either, as if the save on the blobstore fails for some reason you might loose the data, as you will not save it again with the id being in the list

Reply:
>  Risk of concurrency issues here? If saving a big blob, the time it's being saved, we could still try to save an other one as the blobId has not been stored yet...

In which case the behaviour is the same than today :-)

Reply:
> I don't think we should save the blobId first either, as if the save on the blobstore fails for some reason you might loose the data, as you will not save it again with the id being in the list

Can we handler error when saving on the blobStore. If an error throw we can remove blobId, that already saved?